### ISSUE TITLE
Wrap lines in the results window

### ISSUE THREAD

**Issue by divagant-martian** [nvim-telescope/telescope.nvim#701]
**Is your feature request related to a problem? Please describe.**
Using the builtin picker for lsp diagnostics gives me cut version of the error, ,when I really need to be able to read it completely.

**Describe the solution you'd like**
Ideally, something like what's described in #582, but something as simple as `autocmd User TelescopePreviewerLoaded setlocal wrap` with `TelescopePreviewerLoaded` changed for the appropriate name would be enough

**Describe alternatives you've considered**
At least being able to change the layout for the default picker so that I can see the errors in a longer line. This wouldn't really be enough, but I haven't found how to do this either.

**Additional context**
The amount of info that can be read in this width is very little
![image](https://user-images.githubusercontent.com/26765164/112911129-681d4e80-90ba-11eb-95e0-0588b38639f6.png)

**Comment by Conni2461** [comment]
Could be possible. You could change this from `false` to `true`. But this results in a buggy selection. Also it wouldn't fix the truncate problem because the width is column width is hardcoded for diagnostics here. Maybe we should move the table around and do `file:lnum:col | message`.

CC @fdschmidt93

**Comment by fdschmidt93** [comment]
I personally would not move the table around since file names are optional and not shown in document diagnostics.

The following options can be passed to `lsp_{document,workspace}_diagnostics` to reformat the entries.

- `no_sign`: remove E in case of your screenshot to save one char; highlighting indicates error type and is linked to `LspDiagnosticsDefault{Error, Warning, Hint, Information}`
- `line_width`: increase line width of diagnostic message, possibly at the expense of truncating the filename
- `hide_filename`: remove file name entirely by default (`workspace only`)

As a reference, the relevant code can also be found [here](https://github.com/nvim-telescope/telescope.nvim/blob/aefc8317354426864a361d7b7d84848145d218c1/lua/telescope/make_entry.lua#L857).

More generally, I'd prefer to  add an action that calls `vim.lsp.diagnostic.show_line_diagnostic` to "print" the full diagnostic message for the line as that appears clean and rather straightforward though may include other messages. Possible alternative `actions`: (a) put entry in a floating window, (b) horizontal scrolling for entry with `h`, `l` in normal mode? might be of more general use

`vim.lsp.diagnostic.show_line_diagnostic` does not work well out-of-the-box as the floating window is placed in the top left corner and partially hidden. Lspsaga's `show_line_diagnostic` works a bit better as shown below and cleanly overlays the Telescope floating windows. 

I'm for now unfamiliar with the floating window API, but I guess it hopefully boils down to moving the window post-hoc to where the cursor or entry is :) @Conni2461 if you agree, I'd try to make that properly work and submit a PR.. hopefully with sufficient documentation :)

![diagnostic-overlay](https://user-images.githubusercontent.com/39233597/112981679-d36b2d00-915b-11eb-927d-70f8902f9c08.png)

**Comment by divagant-martian** [comment]
Thanks for the alternatives, for the time being I'll use them to be able to see a bit more. As for a) vs b), probably b would generate less problems for different users? a) sounds really cool but b) would perfectly do the trick. Thank you both for the quick response :+1: 

For reference, I'm in rust, where errors are extremely ~~large~~ descriptive, specially if they reference types (like the ones on the screenshot)

**Comment by divagant-martian** [comment]
After your edit, I actually went with `show_line_diagnostic` and mapped it to some key binding. For me it does the trick. Thanks a lot

**Comment by fdschmidt93** [comment]
For anyone else coming here, the below code is also how could invoke line diagnostics for the current entry above the cursor, forgot to share it with the above :)

```lua
local action_state = require('telescope.actions.state')

local line_diagnostics = function()
    local selected_entry = action_state.get_selected_entry()
    local value = selected_entry['value']
    local bufnr, winid = vim.lsp.diagnostic.show_line_diagnostics({}, value.bufnr, value.lnum-1, nil)
    local w = vim.api.nvim_win_get_width(winid)
    -- workaround nvim_win_get_height wrongly returning 1 when called from prompt 
    local h = vim.tbl_count(vim.api.nvim_buf_get_lines(bufnr, 0, -1, false)) 
    local prompt_row, prompt_col = unpack(vim.api.nvim_win_get_position(0))
    vim.api.nvim_win_set_config(winid, {relative='editor', row=prompt_row-h, col=0+prompt_col, width=w, height=h})
end
```

**Comment by divagant-martian** [comment]
this is really nice! How do you achieve this? Slightly off topic, but I' haven't been able to set a binding for (for example) `entry_line_diagnostics`. I'm trying with 
```lua
require('telescope.builtin').lsp_workspace_diagnostics({ 
attach_mappings = function(prompt_bufnr)
    action_set.select:enhance({
      post = entry_line_diagnostics,
    })
    return true
  end,
})
```
Could you share your mapping for this?

**Comment by elianiva** [comment]
toggleable floating window looks very neat! I like that idea :)

**Comment by tami5** [comment]
Oh wow 🤩, I talked to @tjdevries sometime back, it would be nice to reuse it to toggle preview in drop down theme

**Comment by fdschmidt93** [comment]
@divagant-martian, you'd define the mapping as below, where `C-m` is your mapping. Note, the floating window for now will only close with any of these autocommand events "CursorMoved", "CursorMovedI", "BufHidden", "BufLeave", of which none invokes scrolling through the results list. I have yet to think of something neat. The floating window is too raw for the time being, ideally a PR will come soon enough :)

```lua
'telescope.builtin'.lsp_workspace_diagnostics{
  attach_mappings = function(_, map)
    map('i', '<C-m>', entry_line_diagnostics)
    return true
  end
}
```
@Conni2461 @elianiva @tami5 

I'm currently thinking to decompose it into two main functions, a provider that creates floating window and a post-processor that reformats/-sizes, if required, and positions the floating window accordingly (above, below, etc. to fit properly into results). 
Specified like that hopefully makes it as simple as possible to provide new toggleable windows for selected entries for a picker, i.e. by pretty much only having to pass the provider function.
Further, I guess the floating window should be toggleable, i.e. can be closed on repeating the key press, as well as cleared by any movement, e.g. also when the results list is scrolled.
Thoughts?

**Comment by Conni2461** [comment]
I kinda missed the whole conversation :laughing: Sorry
But it seems like this evolved in something similar to this the toggle preview issue. So i just link my thoughts about it: https://github.com/nvim-telescope/telescope.nvim/issues/643#issuecomment-800550342 Maybe that helps :)

**Comment by divagant-martian** [comment]
Encountered the need for wrapped lines again while playing around. When calling `:Telescope command_history` if the command is very long (yes, I am putting whole lua functions there just to play) the command is shown with three dots in the middle, and executed with said dots... producing of course wrong results. I love the idea of being able to toggle a "more info" kind of window, but I think it is very likely that simply wrapping lines is something that people would still need. 

What do you guys think? Should I make another issue for this particular setup?

**Comment by fdschmidt93** [comment]
The issue is that commands are being executed with '...'?

The way commands are executed can be found [here](https://github.com/nvim-telescope/telescope.nvim/blob/d0cf646f65746415294f570ec643ffd0101ca3ab/lua/telescope/builtin/internal.lua#L172). To my knowledge, dots appear due to string formatting and not inherent changes to the set `value`. Are you sure execution of commands with dots is the issue?

On wrapping lines: I'm not familiar with how it's implemented, but wrapped lines would conceptually cause significant overhead as you'd know have to check where lines are wrapped and  change selecting and highlighting etc. adequately which seems like an undertaking while not breaking anything. Using a different layout or toggleable info appears more straightforward and suitable.

**Comment by divagant-martian** [comment]
short video showing the issue with the command history. Typed a long command (what I ended up using to print the whole diagnostic) then `:Telescope command_history` to run it again and I get an error

The only configuration I've done to telescope is using this function as a wrapper for the diagnostics

 `Error loading lua [string ":lua"]:1: '}' expected near '...'`

https://user-images.githubusercontent.com/26765164/113590969-f46acc80-95f8-11eb-846c-14c62c7ffb1d.mp4

**Comment by fdschmidt93** [comment]
Ok, that happens as the `value` is actually truncated if it's longer than the results window since the `entry_maker` is kept more simple for `command_history`. I'll eventually submit a PR that augments the corresponding `make_entry/displayer` such that lines are not truncated in this case. 

That said, if I may, it is probably more sane to experiment with putting a function in a `utils.lua` and map hot reloading of a module to something like in a `globals.lua` ;)

```lua

if pcall(require, 'plenary') then
  RELOAD = require('plenary.reload').reload_module

  R = function(name)
    RELOAD(name)
    return require(name)
  end
end
```

and then everytime you change and save your function in `utils.lua` `R('utils').function()` respects that.

**Comment by zeertzjq** [comment]
The truncating also make it hard to view file names in `lsp_references`. A way to view the full lines will definitely help.

**Comment by Conni2461** [comment]
wrap was just implemented in #1735

Closing, if this isnt to your satisfactory feel free to reopen

**Comment by Masber** [comment]
https://github.com/nvim-telescope/telescope.nvim/pull/1735 does not solve this issue, because diagnostics will still truncate the diagnostic message and wrap the file name instead.

`:Telescope diagnostics theme=dropdown previewer=false wrap_results=true`

![image](https://github.com/user-attachments/assets/c15af04a-aec1-4f36-aba3-e3e4f72173f3)
